zend_abstract_interface/zai_sapi/zai_sapi.c

```diff
@@ -0,0 +1,189 @@
+/* ZAI SAPI is a fork the the embed SAPI:
+ * https://github.com/php/php-src/blob/PHP-8.0.3/sapi/embed/php_embed.c
+ *
+ * We have our own SAPI so that we can provide our own INI entries and custom
+ * SAPI functions for testing. It also gives us more flexibility on adding
+ * entry points for fuzzing. Overall it allows us to more throughly test the
+ * surface area of each component.
+ */
+#include "zai_sapi.h"
+
+#include <main/SAPI.h>
+#include <main/php_main.h>
+#include <main/php_variables.h>
+
+#define UNUSED(x) (void)(x)
+
+static int zai_startup(sapi_module_struct *sapi_module) { return php_module_startup(sapi_module, NULL, 0); }
+
+static int zai_deactivate(void) {
+    fflush(stdout);
+    return SUCCESS;
+}
+
+static inline size_t zai_single_write(const char *str, size_t str_length) {
+#ifdef PHP_WRITE_STDOUT
+    zend_long ret;
+
+    ret = write(STDOUT_FILENO, str, str_length);
+    if (ret <= 0) return 0;
+    return ret;
+#else
+    size_t ret;
+
+    ret = fwrite(str, 1, MIN(str_length, 16384), stdout);
+    return ret;
+#endif
+}
```

Comment:
I know this is copy/paste, but trying to think through it:
1. Taking a quick look, `PHP_WRITE_STDOUT` corresponds to this configure check: https://heap.space/xref/php-src/build/php.m4?r=fa9e2b31#2305. I'm guessing that it may fail on non-Unix platforms like Windows? Dunno.
2. Where is `MIN` defined for these non-Unix platforms?

Reply:
It looks like maybe `write` doesn't buffer, and `fwrite` does: https://stackoverflow.com/a/1360034/538216. So for the CLI sapi at least it may make sense (let me see it asap, don't buffer it). Or something? I don't know.

Reply:
Thanks for calling this out. 

1. I found the [original commit](https://github.com/php/php-src/commit/bc7a582f8480ce92fecde267d4cd3c3f628a4ba9) where that was added back in 2001. I'm not sure about Windows compat.
2. It looks like `MIN` is [part of `zend_portability.h`](https://heap.space/xref/PHP-8.0/Zend/zend_portability.h#417).
3. I think this tries to use `write` because this is supposed to be the "unbuffered write" function for the SAPI. But for our purposes, I think we can just use `fwrite` without any issues.

Reply:
Sammy, is there a "buffered" write operation? If not we may look into the buffering options and just use fwrite :)

Reply:
AFAIK, the SAPIs [only have unbuffered writes](https://heap.space/xref/PHP-8.0/main/SAPI.h#223-224) and all output buffering is handled by [PHP's output buffer](https://www.php.net/manual/en/ref.outcontrol.php) feature. Once PHP is ready to flush it sends it to the SAPI unbuffered write operation.

I updated the [ZAI SAPI unbuffered write operation](https://github.com/DataDog/dd-trace-php/blob/5f22ec49e454c3dbc28596ba962ade4c43a49438/zend_abstract_interface/zai_sapi/zai_sapi.c#L28-L48) to use `fwrite` for the small performance gain. I added this comment:

```
/* This is supposed to be the "unbuffered write" function for the SAPI so using
 * a buffered function like fwrite() here I suppose is technically incorrect.
 * But fwrite()'s buffering gives a little performance boost over write() which
 * better serves the purposes of running many tests through the ZAI SAPI.
 */
```

WDYT?

Reply:
I'm resolving this since the ZAI SAPI has been moved to #1204 and it implements an unbuffered approach to write operations.